src/replica/replica_stub.cpp

```diff
@@ -1136,6 +1135,53 @@ void replica_stub::on_query_app_info(query_app_info_rpc rpc)
     }
 }
 
+// ThreadPool: THREAD_POOL_DEFAULT
+void replica_stub::on_add_new_disk(add_new_disk_rpc rpc)
+{
+    const auto &disk_str = rpc.request().disk_str;
+    auto &resp = rpc.response();
+    resp.err = ERR_OK;
+
+    std::vector<std::string> data_dirs;
+    std::vector<std::string> data_dir_tags;
+    std::string err_msg = "";
+    if (disk_str.empty() ||
+        !replication_options::get_data_dir_and_tag(
+            disk_str, "", "replica", data_dirs, data_dir_tags, err_msg)) {
+        resp.err = ERR_INVALID_PARAMETERS;
+        resp.__set_err_hint(fmt::format("invalid str({}), err_msg: {}", disk_str, err_msg));
+        return;
+    }
+
+    for (auto i = 0; i < data_dir_tags.size(); ++i) {
+        auto dir = data_dirs[i];
+        if (_fs_manager.is_dir_node_available(dir, data_dir_tags[i])) {
+            resp.err = ERR_NODE_ALREADY_EXIST;
+            resp.__set_err_hint(
+                fmt::format("data_dir({}) tag({}) already available", dir, data_dir_tags[i]));
+            return;
+        }
+
+        if (dsn_unlikely(utils::filesystem::directory_exists(dir) &&
+                         !utils::filesystem::is_directory_empty(dir).second)) {
+            resp.err = ERR_DIR_NOT_EMPTY;
+            resp.__set_err_hint(fmt::format("Disk({}) directory is not empty", dir));
+            return;
+        }
+
+        std::string cdir;
+        if (dsn_unlikely(!utils::filesystem::create_directory(dir, cdir, err_msg) ||
+                         !utils::filesystem::check_dir_rw(dir, err_msg))) {
+            resp.err = ERR_FILE_OPERATION_FAILED;
+            resp.__set_err_hint(err_msg);
+            return;
+        }
+
+        ddebug_f("Add a new disk in fs_manager, data_dir={}, tag={}", cdir, data_dir_tags[i]);
```

Comment:
cdir->dir?

Reply:
Both of them is ok, `dir` is from paramter, and `cdir` is the absolute_dir by function `create_directory`.

Reply:
has path: /home/xiaomi/

so `create_directory("/home/xiaomi/pegaus", cdir, err_msg)` return cdir="pegasus" or other?

however, in any case,  I think the log should show the full_path, but absolute path,

Reply:
Acually, the absolute path is full path. 
`create_directory("/home/xiaomi/pegasus", cdir, err_msg)` will get `cdir="/home/xiaomi/pegasus"` 
`create_directory("pegaus", cdir, err_msg)` will also get `cdir="/home/xiaomi/pegasus"` 
The first parameter of `create_directory` is permitted to be relative path.